includes/admin/extensions/class-extensions-api.php

```diff
@@ -0,0 +1,168 @@
+<?php
+
+namespace EDD\Admin\Extensions;
+
+class ExtensionsAPI {
+
+	/**
+	 * Gets the product data from the EDD Products API.
+	 *
+	 * @since 2.11.x
+	 * @param array $body    The body for the API request.
+	 * @param int   $item_id The product ID, if querying a single product.
+	 * @return array|false
+	 */
+	public function get_product_data( $body = array(), $item_id = false ) {
+		if ( empty( $body ) ) {
+			if ( empty( $item_id ) ) {
+				return false;
+			}
+			$body = $this->get_api_body( $item_id );
+		}
+		$key = $this->array_key_first( $body );
+		// The option name is created from the first key/value pair of the API "body".
+		$option_name = sanitize_key( "edd_extension_{$key}_{$body[ $key ]}_data" );
+		$option      = get_option( $option_name );
+		$is_stale    = ! empty( $option['timeout'] ) && time() > $option['timeout'];
+
+		// If the data is "fresh" and what we want exists, return it.
+		if ( ! $is_stale ) {
+			if ( $item_id && ! empty( $option[ $item_id ] ) ) {
+				return $option[ $item_id ];
+			} elseif ( ! empty( $option['timeout'] ) ) {
+				unset( $option['timeout'] );
+
+				return $option;
+			}
+		}
+
+		// The data either does not exist or is expired, so query the API.
+		$url     = add_query_arg(
+			array(
+				'edd_action' => 'extension_data',
+			),
+			$this->get_products_url()
+		);
+		$request = wp_remote_get(
+			$url,
+			array(
+				'timeout'   => 15,
+				'sslverify' => true,
+			)
+		);
+
+		// If there was an API error, set timeout for 1 hour and return stale product data if it exists.
+		if ( is_wp_error( $request ) || ( 200 !== wp_remote_retrieve_response_code( $request ) ) ) {
+			$data = array(
+				'timeout' => strtotime( '+1 hour', time() ),
+			);
+			if ( $option && $is_stale ) {
+				$data = array_merge( $option, array( 'timeout' => strtotime( '+1 hour', $data['timeout'] ) ) );
```

Comment:
Just wondering if this timeout needs to be relative to the original `$data['timeout']` ? Could we just make it `strtotime( '+1 hour', time() )` ?

Potential scenario:

1. Cache expired 1 day ago.
2. User views the page to try again.
3. Request fails.
4. New timeout would be set to: {1 day ago} + 1 hour.
5. Which means it would try again immediately, right?